patroni/validator.py

```diff
@@ -0,0 +1,117 @@
+        exe_file = os.path.join(path, program)
+        if os.path.isfile(exe_file) and os.access(exe_file, os.X_OK):
+            return exe_file
+    return None
+
+def config_validator(config):
+    if "name" not in config:
+        logger.warning("name is not defined")
+    if "scope" not in config:
+        logger.warning("scope is not defined")
+    if "postgresql" not in config:
+        logger.warning("postgresql section is missing")
+    else:
+        if "authentication" not in config["postgresql"]:
+            logger.warning("postgresql.authentication is nor defined")
+        else:
+            if "replication" not in config["postgresql"]["authentication"]:
+                logger.warning("postgresql.authentication.replication is not defined")
+            if "superuser" not in config["postgresql"]["authentication"]:
+                logger.warning("postgresql.authentication.superuser is not defined")
+
+        if "bin_dir" not in config["postgresql"]:
+            logger.info("postgresql.bin_dir is not defined")
+        else:
+            if os.path.exists(config["postgresql"]["bin_dir"]):
+                for program in ["pg_ctl", "initdb", "pg_controldata", "pg_basebackup"]:
+                    exe_file = os.path.join(config["postgresql"]["bin_dir"], program)
+                    if not (os.path.isfile(exe_file) and os.access(exe_file, os.X_OK)):
+                        logger.warning("Program '%s' not found.", program)
+            else:
+                logger.warning("Directory '%s' does not exist.", config['postgresql']['bin_dir'])
+                for program in ["pg_ctl", "initdb", "pg_controldata", "pg_basebackup"]:
+                    if not which(program):
+                        logger.warning("Program '%s' not found.", program)
+        if "data_dir" not in config["postgresql"]:
+            logger.info("postgresql.data_dir is not defined")
+        else:
+            if not data_directory_empty(config["postgresql"]["data_dir"]):
+                if not os.path.exists(
+                    os.path.join(config["postgresql"]["data_dir"], "PG_VERSION")
+                ):
+                    logger.warning(
+                        "data_dir doesn't look like a valid data directory make sure you provide a valid path in configuration"
+                    )
+                elif not os.path.isdir(
+                    os.path.join(config["postgresql"]["data_dir"], "pg_wal")
+                ):
+                    logger.warning(
+                        'data dir for the cluster is not empty, but doesn\'t contain "pg_wal" directory'
+                    )
+            data_dir = config["postgresql"]["data_dir"]
+            if "parameters" not in config["postgresql"]:
+                logger.info("postgresql.parameters is not defined")
+            else:
+                xlogdir = config["postgresql"]["parameters"].get(
+                    "xlogdir", os.path.join(data_dir, "pg_xlog")
```

Comment:
There are no such parameters as waldir or xlogdir.

pg_log and pg_wal are always located in the PGDATA, but they could be symlinks.

Reply:
https://www.postgresql.org/docs/10/app-initdb.html
https://www.postgresql.org/docs/9.6/app-initdb.html

Reply:
You are referring to options of `initdb`, which are not postgres configuration parameters.
There is a way to specify them though, but in a different place:
```yaml
bootstrap:
  initdb:
  - data-checksums
  - waldir: /foo
```

It just tells to initdb that we want the waldir to be located at `/foo` and it will create the symlink `$PGDATA/pg_wal -> /foo`